clang/test/SemaSYCL/sycl-templ-addr-space.cpp

```diff
@@ -0,0 +1,59 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -triple spir64 -verify %s
+
+// Test that an error is issued instead of an assertion for this test case.
+
+template <int b>
+struct c {
+  static constexpr int d = b;
+};
+template <typename ab>
+ab &&l(int);
+// expected-error@+2 {{reference of type 'ah &&' cannot bind to a temporary object because of address space mismatch}}
```

Comment:
I probably don't get this error. Are we sure that there is some address space mismatch?

Reply:
There is definitely an address-space mismatch.   The conversion from (__global ah) to (ah &&) fails.

The full message is something like:
```
sycl-templ-addr-space.cpp:13:9: error: reference of type 'ah &&' cannot bind to a temporary object because of address space mismatch
auto ac(l<ab>(0));
        ^~~~~~~~
...
sycl-templ-addr-space.cpp:57:12: note: while substituting deduced template arguments into function template 'v' [with au = __global ah, $1 = (no value)]
    return pi[0];
           ^
```

Reply:
Yes, I can see address space mismatch on line 57 (but there is also type mismatch, I guess?) and error for this is caught in this test on line 55. But what is wrong with address spaces on line 13 is a bit unclear.

Reply:
The line 13 message is a fallout of the same address-space mismatch.  There was no type-mismatch, one was simply a 'using declaration' of the other.   In any case, with the new change, that message does not get issued.